File: code/python/剑指offer68-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        res = None
        def dfs(node):
            nonlocal res
            if node is None:
                return None
            if node==p:
                if dfs(node.left) is not None or dfs(node.right) is not None:
                    res = node if res is None else res
                return node
            if node==q:
                if dfs(node.left) is not None or dfs(node.right) is not None:
                    res = node if res is None else res
                return node
            getp = dfs(node.left)
            getq = dfs(node.right)
            if getp is not None and getq is not None:
                res = node if res is None else res
                return node
            elif getp is not None or getq is not None:
                return node
            else:
                return None

        dfs(root)
        return res

File: code/java/剑指offer68-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    private TreeNode ans = null;

    public TreeNode dfs(TreeNode node,TreeNode p,TreeNode q){
        if(node==null) return null;
        if(node==p||node==q){
            if(dfs(node.left,p,q)!=null||dfs(node.right,p,q)!=null) ans=ans==null?node:ans;
            return node;
        }
        TreeNode getl = dfs(node.left,p,q);
        TreeNode getr = dfs(node.right,p,q);
        if(getl!=null&&getr!=null){
            ans=ans==null?node:ans;
            return node;
        }else if(getl!=null||getr!=null){
            return node;
        }else{
            return null;
        }
    }

    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        dfs(root,p,q);
        return ans;
    }
}